### PR TITLE
ProcessMonitors API and implementation rework

### PR DESCRIPTION
**Pull request by ben-spiller**
This PR is for a rewrite of the API to make it easier for users to customize and add to monitoring the data gathered on each platform, and to customize how the gathered data is handled (e.g. in case you want to implement automatic leak detection). Careful consideration was given to ways to preserve backwards compatibility but the current API just doesn't permit it. However users who just call startProcessMonitor will find it still behaves the same, so this only affects people who subclasses ProcessMonitor which will be a small minority. 

The new API decouples the collection of data (BaseProcessMonitor) from handlers that do stuff with it (BaseProcessMonitorHandler) such as the ProcessMonitorTextFileHandler that writes data in tsv (or csv) format. 

In case you're wondering, it was necessary to move the OS-specific monitors into a single module because a) otherwise there is nowhere to put common classes shared between them b) the amount of copied+pasted logic and doc between them as a maintenance headache and most importantly c) it allows us to include all OS-specific classes in the doc whcih is useful if you want to subclass them. 

The change that is more likely to affect users is that the default data file format has changed in 3 ways: 
- there is now a (gnuplot-friendly) '#' comment header at the start of the file so you know what the columns mean (could also be machine-read if needed)
- changed the date format which was inconsistent between windows and unix to the standard yyyy-mm-dd format - fixes GH-6. This is a breaking change but worth doing since the previous format inconsistency could have broken tests/tools that read in the file if you happened to run it after the 12th day of a given month
- remove Windows-specific columns so the file is the same on all platforms (and also because I couldn't reliably retrieve the data for them - see below). 

Additional testing uncovered that on linux and windows the results being generated were wildly incorrect in a large number of cases (esp around linux child processes and the fact the windows pdh perf counter api will randomly switch you to monitoring a different pid if an earlier instance of the same .exe basename terminates while you're monitoring). As a result we're going to have to drop the windows-specific keys e.g. thread count, handle count, private bytes as pywin32 doesn't currently support the GetPerformanceInfo function which would get this. PySys is never going to be the best place to write a comprehensive cross-platform process monitoring python library - https://github.com/giampaolo/psutil does that much better. It would be easy using this new PySys API to write a process monitor wrapping psutil which would allow access to a wide array of monitoring statistics on many OSes. I also think we might want to add a Java process monitor at some point that gathers JVM-specific data such as heap/non-heap usage, Java threads, etc. 

Also added ignores= argument to waitForSignal which is rather handy.

### CODE REVIEW

File: pysys/process/monitor.py [hunk 1]
```diff
@@ -0,0 +1,454 @@
...
+		Called on the background monitoring thread regularly.
+		
+		@param sample: An integer starting at 1 and incrementing each time 
+		this method is called. 
+		
+		@return: A dictionary of (typically numeric) values, keyed by 
+		L{ProcessMonitorKey}.
+		@rtype: dict
+		"""
+		raise NotImplementedError('_getData is not implemented yet')
```
Comment (sagscmi) [inline pysys/process/monitor.py]: Maybe a better message would be that "_getData must be implemented in a subclass" or something similar?  The "yet" implies there might one day be an implementation here which I think is unlikely.

File: pysys/process/monitorimpl.py [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+
+			if self._lastValues is not None:
+				if newvalues['time_ns']-self._lastValues['time_ns'] <= 0:
+					# wait a bit longer to avoid div by zero error if the sleeping is somehow messed up
+					self._stopping.wait(min(self.interval, 1))
+					continue
+			
+				lastvalues = self._lastValues
+				break
+			# this is just for the first time _getData is called; need to repeat this once so we have stats to compare to
```
Comment (sagscmi) [inline pysys/process/monitorimpl.py]: Uber-nitpick: add a blank line before the comment :-)
```suggestion
			
			# this is just for the first time _getData is called; need to repeat this once so we have stats to compare to
```

File: pysys/process/monitor.py (hunk 1 above)
Comment (ben-spiller) [reply pysys/process/monitor.py]: yes fixed

**Comment by ben-spiller** [comment]
> Just a few formatting/clarity nits, and a question: would it make sense to allow `DEFAULT_PROCESS_MONITOR` to also be overridden somehow? Or is the recommended way to use a different monitor class for all your tests to override `BaseTest.startProcessMonitor()`?

startProcessMonitor is for starting the default in-the-box one; if you want to start a different one you'd jsut do it by instantiating and calling start() on your own class (startProcessMonitor()'s implementation is deliberately trivial to make that easy), potentially in a helper method if you want it shared across lots of tests. For example if you wrote one to monitor Java JVM stats you might add startJVMProcessMonitor. Or you could of course override BaseTest.startProcessMonitor in a subclass too. I think that's sufficiently easy and sufficiently similar to what pysys users would be doing anything that there's no need to invent a separate mechanism for customizing the "default" process monitor.

File: pysys/process/monitorimpl.py (hunk 2 above)
Comment (ben-spiller) [reply pysys/process/monitorimpl.py]: fixed

**Pull request merged**